### PR TITLE
PHP SDK maintainance

### PR DESCRIPTION
**Pull request by jenschude**
1. code and add tests that cover the created code. Your code should be warning-free.
   * [x] tests existing
1. stick to PSR-2 and and don't reformat existing code.
   * [ ] code style check succesful

### CODE REVIEW

File: tests/unit/Model/RamlModelTest.php [hunk 1]
```diff
@@ -479,6 +479,9 @@ public function getModelClasses($searchPath)
             if (strpos($class, 'Core\\Helper') > 0) {
                 continue;
             }
+            if (strpos($class, 'Core\\Model\\Message') > 0) {
```
Comment (jenschude) [inline tests/unit/Model/RamlModelTest.php:482]: This disables some of the tests regarding messages. The correct way would be to add the failing message definition to the api specs

Comment (barbara79) [reply tests/unit/Model/RamlModelTest.php]: I didn't realize that the Message didn't exist in the API specs ;) . I just open the PR with the changes

**Comment by barbara79** [comment]
this PR has a dependency on https://github.com/commercetools/commercetools-docs/pull/1506.
The tests will still fail until this PR will be merged

**Comment by barbara79** [comment]
> this PR has a dependency on [commercetools/commercetools-docs#1506](https://github.com/commercetools/commercetools-docs/pull/1506). The tests will still fail until this PR will be merged

It has been merged 👍

File: src/Core/Request/Customers/CustomerLoginRequest.php [hunk 2]
```diff
@@ -67,16 +69,19 @@ class CustomerLoginRequest extends AbstractApiRequest
     /**
      * @param string $email
      * @param string $password
-     * @param string $anonymousCartId
+     * @param string $anonymousCart
```
Comment (jenschude) [inline src/Core/Request/Customers/CustomerLoginRequest.php:73]: please update the param here

```suggestion
     * @param CartReference|string $anonymousCart
```

File: src/Core/Request/Customers/CustomerLoginRequest.php [hunk 3]
```diff
@@ -67,16 +69,19 @@ class CustomerLoginRequest extends AbstractApiRequest
...
     {
         parent::__construct(LoginEndpoint::endpoint(), $context);
         $this->email = $email;
         $this->password = $password;
-        $this->anonymousCartId = $anonymousCartId;
-        $this->anonymousCart = $anonymousCart;
+        if ($anonymousCart instanceof CartReference) {
+            $this->anonymousCart = $anonymousCart;
+        } else {
+            $this->anonymousCartId = $anonymousCart;
```
Comment (jenschude) [inline src/Core/Request/Customers/CustomerLoginRequest.php:84]: You could create a CartReference here

```suggestion
            $this->anonymousCart = CartReference::ofId($anonymousCart);
```

File: src/Core/Request/Customers/CustomerLoginRequest.php [hunk 4]
```diff
@@ -197,31 +202,31 @@ public function setAnonymousCart($anonymousCart)
...
+     * @param CartReference|null $anonymousCart
      * @param Context $context
      * @return static
      */
     public static function ofEmailPasswordAndUpdateProductData(
         $email,
         $password,
         $updateProductData,
-        $anonymousCartId = null,
+        CartReference $anonymousCart = null,
```
Comment (jenschude) [inline src/Core/Request/Customers/CustomerLoginRequest.php]: Breaking change as you limit it now to CartReference only

Btw the setAnonymousCartId method should set the anonymousCart field to null as both fields are exclusive or connected

File: src/Core/Request/Orders/OrderCreateFromCartRequest.php [hunk 5]
```diff
@@ -56,15 +56,19 @@ class OrderCreateFromCartRequest extends AbstractApiRequest
...
-    public function __construct($cartId = null, CartReference $cart = null, $version = null, Context $context = null)
+    public function __construct($cart, $version = null, Context $context = null)
     {
         parent::__construct(OrdersEndpoint::endpoint(), $context);
-        $this->setCartId($cartId)->setCart($cart)->setVersion($version);
+
+        if ($cart instanceof CartReference) {
+            $this->setCart($cart)->setVersion($version);
+        } else {
+            $this->setCartId($cart)->setVersion($version);
```
Comment (jenschude) [inline src/Core/Request/Orders/OrderCreateFromCartRequest.php:70]:
```suggestion
            $this->setCart(CartReference::ofI($cart))->setVersion($version);
```

File: src/Core/Request/Orders/OrderCreateFromCartRequest.php [hunk 6]
```diff
@@ -182,25 +226,26 @@ public function setShipmentState($shipmentState)
...
 
     /**
-     * @param string $cartId
+     * @param CartReference $cart
      * @param int $version
      * @param Context $context
      * @return static
      */
-    public static function ofCartIdAndVersion($cartId, $version, Context $context = null)
+    public static function ofCartAndVersion(CartReference $cart, $version, Context $context = null)
```
Comment (jenschude) [inline src/Core/Request/Orders/OrderCreateFromCartRequest.php:246]: This is a breaking change

File: src/Core/Request/Orders/OrderCreateFromCartRequest.php [hunk 7]
```diff
@@ -220,7 +265,7 @@ public function buildResponse(ResponseInterface $response)
     public function httpRequest()
     {
         $payload = [
-            static::ID => $this->getCartId(),
+            static::CART => $this->getCart(),
```
Comment (jenschude) [inline src/Core/Request/Orders/OrderCreateFromCartRequest.php:268]: Old code will stop working as the ID is not transmitted anymore.

So all setters/getters should set the cart field and return the ID from the cart field

File: src/Core/Request/Customers/CustomerLoginRequest.php (hunk 2 above)
Comment (barbara79) [reply src/Core/Request/Customers/CustomerLoginRequest.php:73]: yes, I didn't see it ;)

File: src/Core/Request/Customers/CustomerLoginRequest.php (hunk 4 above)
Comment (barbara79) [reply src/Core/Request/Customers/CustomerLoginRequest.php]: I don't know why it didn't update this from my locale I put CartReference|string as type hinting

**Pull request merged**